packages/cloudflare/src/api/kvCache.ts

```diff
@@ -0,0 +1,150 @@
+
+import { getCloudflareContext } from "./get-cloudflare-context.js";
+
+export const CACHE_ASSET_DIR = "cnd-cgi/_next_cache";
+
+export const STATUS_DELETED = 1;
+
+/**
+ * Open Next cache based on cloudflare KV and Assets.
+ *
+ * Note: The class is instantiated outside of the request context.
+ * The cloudflare context and process.env are not initialzed yet
+ * when the constructor is called.
+ */
+class Cache implements IncrementalCache {
+  readonly name = "cloudflare-kv";
+  protected initialized = false;
+  protected kv: KVNamespace | undefined;
+  protected assets: Fetcher | undefined;
+
+  async get<IsFetch extends boolean = false>(
+    key: string,
+    isFetch?: IsFetch
+  ): Promise<WithLastModified<CacheValue<IsFetch>>> {
+    if (!this.initialized) {
+      await this.init();
+    }
+
+    if (!(this.kv || this.assets)) {
+      throw new IgnorableError(`No KVNamespace nor Fetcher`);
+    }
+
+    this.debug(`Get ${key}`);
+
+    try {
+      let entry: {
+        value?: CacheValue<IsFetch>;
+        lastModified?: number;
+        status?: number;
+      } | null = null;
+
+      if (this.kv) {
+        this.debug(`- From KV`);
+        const kvKey = this.getKVKey(key, isFetch);
+        entry = await this.kv.get(kvKey, "json");
+        if (entry?.status === STATUS_DELETED) {
+          return {};
+        }
+      }
+
+      if (!entry && this.assets) {
+        this.debug(`- From Assets`);
+        const url = this.getAssetUrl(key, isFetch);
+        const response = await this.assets.fetch(url);
+        if (response.ok) {
+          // TODO: consider populating KV with the asset value if faster.
+          // This could be optional as KV writes are $$.
+          // See https://github.com/opennextjs/opennextjs-cloudflare/pull/194#discussion_r1893166026
+          entry = {
+            value: await response.json(),
+            // __BUILD_TIMESTAMP_MS__ is injected by ESBuild.
+            lastModified: (globalThis as { __BUILD_TIMESTAMP_MS__?: number }).__BUILD_TIMESTAMP_MS__,
+          };
+        }
+      }
+      this.debug(entry ? `-> hit` : `-> miss`);
+      return { value: entry?.value, lastModified: entry?.lastModified };
+    } catch {
+      throw new RecoverableError(`Failed to get cache [${key}]`);
+    }
+  }
+
+  async set<IsFetch extends boolean = false>(
+    key: string,
+    value: CacheValue<IsFetch>,
+    isFetch?: IsFetch
+  ): Promise<void> {
+    if (!this.initialized) {
+      await this.init();
+    }
+    if (!this.kv) {
+      throw new IgnorableError(`No KVNamespace`);
+    }
+    this.debug(`Set ${key}`);
+    try {
+      const kvKey = this.getKVKey(key, isFetch);
+      // Note: We can not set a TTL as we might fallback to assets,
+      //       still removing old data (old BUILD_ID) could help avoiding
+      //       the cache growing too big.
+      await this.kv.put(
+        kvKey,
+        JSON.stringify({
+          value,
+          lastModified: Date.now(),
```

Comment:
Correct me if i'm wrong, but isn't `Date.now()` the time of the last IO in workers ?
Maybe you should add a comment about this here, because ISR with low revalidation time might be affected by this. It probably doesn't matter in like 99% of the cases, but it's worth mentioning

Reply:
Great idea, added a comment.
Thanks!